### DISCUSSION TITLE
How to add sqlcmd into ARM azure-sql-edge?

### DISCUSSION THREAD

**Discussion by d4vidsha** [microsoft/go-sqlcmd#501] [Q&A] [answered]
Our setup is similar to what is shown [here](https://github.com/microsoft/mssql-docker/tree/master/linux/preview/examples/mssql-customize) and works for x86 processors only.

But we would like to extend this Docker implementation so that it works for ARM-based processors. The build will succeed, but since [SQL client tools aren't available on the ARM64 platform](https://learn.microsoft.com/en-us/azure/azure-sql-edge/connect), we are stuck and unable to run sqlcmd to initialize our database. The obvious error we get in Docker is:
```
/opt/mssql-tools/bin/sqlcmd: No such file or directory
```
which is supposedly coming from [this line of code](https://github.com/microsoft/mssql-docker/blob/80e2a51d0eb1693f2de014fb26d4a414f5a5add5/linux/preview/examples/mssql-customize/configure-db.sh#L14C13-L14C40).

We are relatively new to MSSQL and Docker, so my simple question is, how can I use go-sqlcmd so that it replaces the one from `mssql-tools`? Not sure if this is the right approach, but I got the idea from [this StackOverflow post](https://stackoverflow.com/a/75133627/15444163) as well as a previous discussion post https://github.com/microsoft/go-sqlcmd/discussions/36#discussioncomment-2192042.

Additional resources
---
- https://learn.microsoft.com/en-gb/sql/tools/sqlcmd/sqlcmd-utility
- https://hub.docker.com/_/microsoft-azure-sql-edge

**Comment by apoorvdeshmukh** [comment]
We regularly release arm64 binary of sqlcmd under [releases](https://github.com/microsoft/go-sqlcmd/releases/latest).  
As a part of your script, could you try fetching [this binary](https://github.com/microsoft/go-sqlcmd/releases/download/v1.5.0/sqlcmd-v1.5.0-linux-arm64.tar.bz2) and run the intended sqlcmd command? 
You may have to change your code to ensure that you are invoking the downloaded sqlcmd instead of the one found under /opt/mssql-tools/

**Reply by d4vidsha** [reply]
Cheers @apoorvdeshmukh 👍

**Comment by d4vidsha** [comment, marked as answer]
After some further digging, I'm downloading the go-sqlcmd binaries directly into our `Dockerfile`, utilising the convention outlined [here](https://github.com/BretFisher/multi-platform-docker-build?tab=readme-ov-file#add-dockerfile-logic-to-detect-the-platform-it-needs-to-use).

An example `Dockerfile`:
```
FROM mcr.microsoft.com/azure-sql-edge:latest

USER root

# add sqlcmd from https://github.com/microsoft/go-sqlcmd
# we add this to /opt/mssql-tools/bin which completely overrides the
# sqlcmd from the base image. Read up on why we do this here:
# https://github.com/microsoft/go-sqlcmd/discussions/501#discussion-6088877
WORKDIR /opt/mssql-tools/bin
ENV GOSQLCMD_VERSION=v1.5.0
ARG TARGETPLATFORM
RUN case ${TARGETPLATFORM} in \
    "linux/amd64")  GOSQLCMD_ARCH=amd64 ;; \
    "linux/arm64")  GOSQLCMD_ARCH=arm64 ;; \
    *) echo "Unsupported platform: ${TARGETPLATFORM}"; exit 1 ;; \
    esac \
 && wget https://github.com/microsoft/go-sqlcmd/releases/download/${GOSQLCMD_VERSION}/sqlcmd-${GOSQLCMD_VERSION}-linux-${GOSQLCMD_ARCH}.tar.bz2 \
 && tar -xjf sqlcmd-${GOSQLCMD_VERSION}-linux-${GOSQLCMD_ARCH}.tar.bz2

# now we set up the usual
WORKDIR /app
COPY setup.sql setup.sql
COPY entrypoint.sh entrypoint.sh
COPY init_db.sh init_db.sh
RUN chmod +x entrypoint.sh
RUN chmod +x init_db.sh

CMD ./entrypoint.sh
```
